### PR TITLE
Add -depfile option to save dependency info

### PR DESCRIPTION
**Pull request by apanteleev**
This change adds a `-depfile` command line option to `slangc` to save the dependency information, i.e. which source files does every output of the current compile request depend on. The deps are saved in a regular Makefile-type format, similar to e.g. `glslangValidator`.

### CODE REVIEW

File: source/slang/slang-compiler.cpp [hunk 1]
```diff
@@ -2177,6 +2177,103 @@ namespace Slang
...
+            case '[':
+            case ']':
+            case '\\':
+                builder.appendChar('\\');
+                break;
+
+            case '$':
+                builder.appendChar('$');
+                break;
+            }
```
Comment (jsmall-nvidia) [inline source/slang/slang-compiler.cpp:2205]: Is the '$' case necessary?

File: source/slang/slang-compiler.cpp [hunk 2]
```diff
@@ -2177,6 +2177,103 @@ namespace Slang
         return SLANG_OK;
     }
 
+    static void writeString(Stream& stream, const char* string)
+    {
+        stream.write(string, strlen(string));
+    }
+
+    static void escapeDependencyString(const char* string, StringBuilder& builder)
```
Comment (jsmall-nvidia) [inline source/slang/slang-compiler.cpp:2185]: Nitpik: For variables that are 'output' we have a convention of prefixing with `out` so `outBuilder` here. If it's in/out we prefix with `io`.

File: source/slang/slang-compiler.cpp [hunk 3]
```diff
@@ -2177,6 +2177,103 @@ namespace Slang
...
+                break;
+            }
+
+            builder.appendChar(c);
+        }
+    }
+
+    // Writes a line to the file stream, formatted like this:
+    //   <output-file>: <dependency-file> <dependency-file...>
+    static void writeDependencyStatement(Stream& stream, EndToEndCompileRequest* compileRequest, const String& outputPath)
```
Comment (jsmall-nvidia) [inline source/slang/slang-compiler.cpp:2213]: Nitpik: By convention (although certainly not always used by any means) for functions that are local to a .cpp file (or are not part of the public interface of a class) we prefix with _. So `_writeDependencyStatement` here.

File: source/slang/slang-compiler.cpp (hunk 1 above)
Comment (apanteleev) [reply source/slang/slang-compiler.cpp:2205]: Yes: per Makefile conventions, `$` is escaped as `$$`, unlike other characters.

File: source/slang/slang-compiler.cpp (hunk 2 above)
Comment (apanteleev) [reply source/slang/slang-compiler.cpp]: Done.

File: source/slang/slang-compiler.cpp (hunk 3 above)
Comment (apanteleev) [reply source/slang/slang-compiler.cpp]: Done.

File: source/slang/slang-compiler.cpp (hunk 1 above)
Comment (jsmall-nvidia) [reply source/slang/slang-compiler.cpp:2205]: Ok makes sense.

**Review by jsmall-nvidia — APPROVED** [review]

**Review by jsmall-nvidia — APPROVED** [review]

**Pull request merged**